Clever account takeover

ewjordan opened the pull request:
Gives users logging in for the first time via Clever the option to attach to a pre-existing Code.org account:

<img width="1026" alt="screen shot 2017-10-09 at 12 16 16 pm" src="https://user-images.githubusercontent.com/70630/31354634-dbd0ea32-aceb-11e7-9e50-746802f45bee.png">

If the user picks the "link account" option, they are directed to a login page; when they log in, the Clever account will take over as the oauth provider of the existing account without changing any of the other details (and the original Clever-linked Code.org account which is auto-created when signing in via Clever is deleted).

This is mostly plumbing to pass the necessary oauth parameters around as well as triggers to tweak the way the sign in and sign up pages behave.

This behavior is behind a pagemode flag (activated via /private/pagemode/clever_takeover) because it touches a lot of very critical code and needs some manual bashing.

ewjordan commented on dashboard/app/models/user.rb:
```diff
@@ -605,7 +605,7 @@ def self.name_from_omniauth(raw_name)
   CLEVER_ADMIN_USER_TYPES = ['district_admin', 'school_admin'].freeze
   def self.from_omniauth(auth, params)
-    omniauth_user = where(provider: auth.provider, uid: auth.uid).first_or_create do |user|
+    omniauth_user = find_or_create_by(provider: auth.provider, uid: auth.uid) do |user|
```
This might not be necessary for this particular feature, and could potentially be dangerous, so I might pull this out.

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+    fontSize: 15,
+    lineHeight: '22px'
+  },
+  highlightRow: {
```
`styles.highlightRow` seems to be unused.  I'm guessing this style block was copy-pasted from somewhere - can we deduplicate this?

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+  hr: {
+    borderColor: color.light_gray
+  },
+  email: {
```
Likewise `styles.email` looks unused.

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+ * become visible on the accounts page after the transition, which is a
+ * potential violation of student privacy).
+ */
+export default class ConfirmEmailModal extends React.Component {
```
The class name and doc comment above are a mismatch for the filename; I assume this should be `LinkCleverAccountModal`. 

(Turns out this doesn't break anything because it's the default export and the name is unused, but it might have an impact later in debugging or testing).

ewjordan replied:
Yup, that's not good, good catch!

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+  constructor(props) {
+    super(props);
+    this.state = {};
+  }
```
Since we have the class-property syntax extension enabled you can replace the whole constructor here with
```jsx
state = {};
```

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+    this.props.handleSubmit();
+  };
+
+  render = () => {
```
Nit: There's no need to make `render` an arrow function, React will always call it in the context of the component.

islemaster commented on apps/src/code-studio/LinkCleverAccountModal.jsx:
```diff
@@ -0,0 +1,126 @@
+          </button>
+          <button
+            onClick={this.ok}
+            style={Object.assign({},styles.buttonPrimary)}
```
Nit: Can just pass `styles.buttonPrimary` here, the `button` component isn't going to modify that object.

islemaster commented on apps/src/sites/studio/pages/home/_homepage.js:
```diff
@@ -165,3 +166,39 @@ function showHomepage() {
 }
+
+window.CleverTakeoverManager = function (options) {
+  this.options = options;
```
Nit: You're closed over the `options` argument here, so you can omit this assignment and just refer to `options` any place you are using `this.options` or `self.options`.  You won't have to think about `this` binding at all.

It might be even more expressive to do a destructing assignment right in the function parameter list, which would make explicit the available options and let the rest of your code get even more compact.

```js
window.CleverTakeoverManager = function ({
  cleverLinkFlag,
  userIDToMerge,
  mergeAuthToken
}) {
  // ...rest of the module...

  // Inside onConfirmLink:
  window.location.href = `/users/clever_takeover?mergeID=${userIDToMerge}&token=${mergeAuthToken}`;
};
```

islemaster commented on apps/src/sites/studio/pages/home/_homepage.js:
```diff
@@ -165,3 +166,39 @@ function showHomepage() {
+  }
+
+  function onConfirmLink() {
+    window.location.href = "/users/clever_takeover?mergeID=" + self.options.userIDToMerge + "&token=" + self.options.mergeAuthToken;
```
Is the `mergeAuthToken` safe to send to the client?

ewjordan replied:
I'm going to do a security review on this before pushing live, because there are some other things I worry a lot about with this flow.

islemaster commented on apps/src/sites/studio/pages/home/_homepage.js:
```diff
@@ -165,3 +166,39 @@ function showHomepage() {
+    showLinkCleverModal(onCancelModal, onConfirmLink);
+  }
+
+  function closeLinkCleverModal() {
```
Nit: This one-line function is only used in one place - I'd just move the body of this function to `onCancelModal`.

islemaster commented on dashboard/app/helpers/users_helper.rb:
```diff
@@ -5,6 +5,22 @@ module UsersHelper
   include ApplicationHelper
   include SharedConstants
 
+  def check_and_apply_clever_takeover(user)
```
Should this method check `cookies['pm'] == 'clever_takeover'` and raise if it's reached without the flag, just in case?

ewjordan replied:
:+1:

islemaster commented on dashboard/app/views/home/_homepage.html.haml:
```diff
@@ -60,3 +60,13 @@
+      userIDToMerge: "#{session['clever_takeover_id']}",
+      mergeAuthToken: "#{session['clever_takeover_token']}",
+    });
+  });
```
I think it would be preferable to move this `$(document).ready` handler into `_homepage.js` as well; and you wouldn't have to put `CleverTakeoverManager` on window anymore.  Pass the session variables here to the script via the `homepage_data` hash, above.

islemaster commented on dashboard/app/controllers/sessions_controller.rb:
```diff
@@ -21,9 +23,25 @@ def new
     super
   end
 
+  # POST /resource/clever_takeover
```
Should this be `POST /users/clever_takeover`?

ewjordan replied:
I was following the pattern elsewhere in the file here, you're probably right that these should all be changed. Will hit in cleanup.

ewjordan commented:
Note: merged temporarily without addressing all feedback for bug bashing purposes, may revert if anything comes up, otherwise address feedback in a follow-up PR.
